README.md

# wingpod_proc

**wingpod_proc** is a small Python package to process raw data logged by the [**hgpstools**](https://bitbucket.org/haukex/hgpstools/src/master/) package within the WINGPOD turbulence measuring system. It was designed (at least attempted) to be as simple and compact as possible using [**pandas'**](https://pandas.pydata.org) [**DataFrame**](https://pandas.pydata.org/pandas-docs/stable/dsintro.html) objects.

Key features are:

* calculation of wind components in an earth-bounded cartesian coordinate system
* export of processed data as csv and kmz
* provision of coefficients calibration routines as [**Jupyter**](http://jupyter.org/) notebooks (needed to calculate wind components accurately)
* extensive documentation (e.g. master thesis in german relating to the physics background)

___

## Installation

### Requirements
#### Software

Make sure **Python 3.6** or greater is installed. Older versions may work, but have not been tested.

Before you proceed, make sure the following libraries are installed, too,:

* numpy (v1.15.2 or greater)
* pandas (v0.23.4 or greater)
* matplotlib (library not used yet, will be added soon)
* simplekml (v1.2.8 or greater)
* zipfile (version unknown, may be added)

by hitting `import [name of desired library without braces]` into a **Python console** to import them. If a `ModuleNotFoundError` is returned, the desired library is missing and has to be installed. In this case, proceed by running `pip install [library]` in a **terminal** of your choice (e.g. Bash Shell) and try to import the library in Python again.

After you have imported the libraries listed above, check their respective versions by prompting `[library].__version__` in the **Python console** (does not work with *zipfile*!).

#### Hardware

**LOTS of RAM** (16 GB or more recommended).

There's an issue of allocating huge amounts of RAM during the processing, which (hopefully) will be addressed as soon as possible.

### Actual Installation

Just clone the repository wherever you want it to be.

To guarantee consistent and reproducible results, it is **recommended** to run the `test.py` script by prompting `python test.py` in a console after installation. A couple of `passed!` or `failed!` messages will show up, indicating if the results of the testdata processing is consistent with reference values stored in the `/testdata/testvar.py` file.

___

## Usage

To calculate the wind components straight from a given set of input log files, go into the main directory of the package and open `conf.py` with a text editor of your choice. It is **mandatory** to set three paths according to your setup:

* `data_dir` - path in which the raw input files are stored.
* `mod_dir` - path into which wingpod_proc package was cloned to.
* `out_dir` - output path for csv, kmz and plot files.

All other variables are optional and explained in the `conf.py` in detail.

**To run the package, simply hit `python main.py` in a terminal.**

If you encounter **serious memory issues** (system freezing), try to decrease `upsmpl_rate` or chop down the input files into smaller units.
___

## Description of package directories and modules

### Main directory

There are four files included in the main directory, which are the most relevant from a user's point of view:

* `main.py` - contains main work flow, loads and calls subroutines.
* `conf.py` - configuration file, explained above.
* `const.py` - constant file, contains a bunch of correction parameters retrieved by calibration routines.
* `test.py` - test routine, explained above.

While the `conf.py` file has to be frequently edited by the user according to his needs, the `const.py` includes also some important features which have to be addressed from time to time:

1. calibration constants should be reviewed and evaluated regulary (at least once in a year) by performing the calibration maneuvers and calculate the respective parameters using the calibration routines.
2. Difference between TAI and UTC (leap seconds) have to be [checked](https://hpiers.obspm.fr/iers/bul/bulc/bulletinc.dat) and set in respect to the time frame of the data to be processed. For instance, TAI/UTC-difference is 37 s for all raw data recorded in 2017 and 2018. This is necessary to calculate the exact GNSS time (does not account for leap seconds) received by the SPAN-IGM as time reference into UTC based UNIX Epoch timestamps (does account for leap seconds) of the output data.

### Subdirectories

The following subdirectories are included:

* `calibration/` - calibration routines (Jupyter notebooks) and a bunch of config./const. files.
* `data/` - wind barb png's for kmz export and X-Plane test data (latter not in use).
* `doc/` - master thesis and presentation in german which cover the physical/technical background.
* `lib/` - several python functions. They are organized according their purpose (loading data, calculating wind, exporting etc.).
* `script_graveyard/` - old scripts went in there, not in use.
* `testdata/` - test data for test.py routine.
* `tools/` - parser tool to split concatenated raw data files by looking for logger START/STOP statements (Bash shell only).
___

## Known issues and To-Do's (maybe...):

* Calibration routines are not automated.
* Way to much RAM gets allocated, if a high upsampling rate (upsmpl_rate >100) was selected or the raw data files are quite extensive (several hours of data). No investigation has been done so far in this regard.
* Comma seperated output file is HUGE and compression takes ages. At least a formatted output would be nice to save as much symbols as possible.
* Some sort of verbose output would be nice (processing time, mean values for wind components, plots, packages versions etc...).

___

### Author
*Alexander B., ka42001@gmx.net*


changelog.txt
[2019/02/10]
Changes for last commit of this date:

- updated static defect calibration routines

[2019/01/24]
Changes for last commit of this date:

- changed directory handling: raw input files have to be put
	into the 'in/' directory. Output will be saved into
	the 'out/' directory.

- removed unused routines in flow angle calibration guide,
	still not finished

[2018/12/10]
Changes for last commit of this date:

- added .tex sources for calibration guide and presentation
- added a_priori_const.py file in calibration subdir, used
	as 'starting point' for the whole calibration process

- minor change in calibration guide (calibration-square parallel
	to main cardinal directions)
- revised static defect calibration routines

- removed unused const.py files in calibration subdir

[2018/11/26]
Changes for last commit of this date:

- added documentation (README and thesis)
- added calibration routines (not tested yet!)

- removed some unused things

[2018/10/12]
Changes for last commit of this date:

- minor fixes regarding the test routine

[2018/10/09]
Changes for last commit of this date:

- Updated test procedure so that the user can check
	the calculation results for consistency 

[2018/10/08]
Changes for last commit of this date:

- fixed a bug which occured in pandas v0.23.4

[2018/08/27]
Changes for last commit of this date:

- added constant to compensate dynamic pressure errors

- updated coefficients

[2018/08/20]
Changes for last commit of this date:

- added INS status legend in exported kml file

- changed (increased) pins density in kml file

[2018/08/15]
Changes for last commit of this date:

- fixed wind barbs showing wrong wind directions (180 deg) in exported
	kmz file
- fixed temperature and dewpoint units in exported kmz file.

- added file parser for raw data files. Pressure data files can be plotted
	to decide with ease, which files may contain actual flight data.

[2018/06/14]
Changes since last commit:

- fixed small bug in the master timestamp generation. the desired
	upsampled frequency was not equal as specified in the configuration.
- fixed missing old main.py, which wasn't updated in the last commit

[2018/06/13]
Changes since last commit:

- added calculation of meteorological values (potential temperature,
	equivalent potential temperature (also known as Theta_e),
	virtual temperature, air density)
- added calculation of indicated airspeed
- added SI converation just after raw data loading to eliminate the
	SI/non-SI mess. Configurable constants where left in userfriendly
	non-SI values, though (for example: temperature in °C instead of K).

- removed unused load_mevis function in the load_data.py collection

The output was tested for consistency using the proposed testdata

Known issues/to do:
- split unix timestamps somehow. After upsampling, the timestamp values are
	so big that they suffer precision losses because of limited float
	precision. This causes a very small but disturbing artificial jitter
	in the output timestamp values. One way to address this problem is to
	split the UNIX time of the reference day at 00:00 UTC and the seconds
	thereafter.

[2018/06/10]
Changes since last commit:

- added testdata files and description, contained in the ./testdata directory

[2018/05/25]
Changes since last commit:

- fixed turning rates and acceleration values which were in rad/s/sample and
  	m/s/sample for the internal IMU sampling rate (125 Hz)
- fixed sign for angular velocities
- finally added the correction constants obtained from calibration maneuvers

- some tiny syntax fixes here and there

This release is presumed to be the final version to use the windpod-proc
	package operationally.
	However, the results were'nt discussed very profoundly
	yet, therefore, the correction algorithm might change in the
	future. There is still some ucertainty, if the (measureable) static
	defect caused by flow angle deflection should be properly corrected
	and implemented.

[2018/04/23]
Changes since last commit:

- fixed path to the wind-barbs directory
- fixed and tested csv-export for processed output-data. Takes ages...but works.
- fixed the unintentionally deactivated inclusion of boom lenght in the wind
  	calculation

- added offset correction for pressure sensores in order to re-zero them
- added dynamic correction for static and dynamic pressure measurements
- added response time offset for pressure sensors
- added time range selection option in the configuration
- added output directory option for data exports

- removed AD conversation lag subtraction of pressure sensor timestamps

- some comments have been revised
- code is more consistently regarding code convetions. Not 100% hardcore
       pythonian, though.


Known issues:
- jumps in timestamps can be problematic and leed to errors in the preprocessing
  	part. This can be bypassed manually by removing timestamp jumps in raw
	data files. Keeping a consistant timestamp in raw data files saves time
	consuming manupulation of them.
- huge amounts of RAM get allocated, package may not be usable on machines with
	low end memory space (4 GB or less).


lib/load_data.py
#Loading routines for Wingpod ASCII data streams and MEVIS csv files with
#   pressure data
#2018/06/13
#Alexander Buetow
#Institute for Space Sciences
#Free University of Berlin

#+++MODULES+++
import numpy as np
import re
import pandas as pd
from datetime import datetime

#+++FUNCTIONS+++
def skiprow_column_finder(file, sep=None):
    """
    Finds an returns unusable rows without a timestamp, erroneous lines and
    status prints which are useless for the postprocessing. Also returns the
    max number of columns of a csv file which is usable for the User
    
    Input:
        file - file string
        sep - seperator(s) of columns in csv file
        
    Output:
        rows - rows to be skipped
        max_n - maximum number of columns
    """       
    f = open(file, "r")
    n = 1
    rows = list()
    max_n=0
    for line in f:
        num = line.rstrip()[0:17]
        try:
            num = float(num) #fine, if first characers is a number -> timestamp
        except:
            rows.append(n) #if not, append the line number
        if not sep == None:
            words = len(re.split(sep, line)) #seperate line to count columns
            if words > max_n:
                max_n = words
        n += 1
        
    if sep == None:
        return(np.array(rows)-1)
    else:
        return(np.array(rows)-1,max_n)
        

def dataframe_to_numeric(df, cols=None):
    """
    Converts pd.DataFrame columns to numeric
    
    Input:
        df - pd.DataFrame
        cols - list of column names to convert
        
    Output:
        df - pd.DataFrame with numeric cols
    """
    if cols == None:
        cols = df.columns
    for i in cols:
        df[i] = pd.to_numeric(df[i], errors='coerce')
    df = df.dropna()
    
    return(df)
    

def load_cpt6100(file, col, delim='\ '):
    """
    reads in pressure data from cpt6100 data files
    
    Input:
        file - str, file string
        col - str, name of pressure value column
        delim - str, regex of delimiter
        
    Output:
        df - plain data from file
    """
    print('Loading CPT6100 data...')
    t = skiprow_column_finder(file)
    df = pd.read_csv(file, skiprows=t, delim_whitespace=True, names=[
            'time', 'x', col])
    df = df.drop(['x'], axis=1) #checksum column not needed
    df = df.dropna()
    df[col] = df[col] * 100 #output in mbar or hPa, calculation into Pa
    
    return(df)

def load_hmt310(file):
    """
    Reads in temperature and humidity data from HMT310 data files.
    
    Input:
        file - file string
        delim - delimiter
        
    Ouput:
        data - plain data from file
    """
    
    print('Loading HMT310 data...')
    t = skiprow_column_finder(file)
    time, rh, t, td, a, x, pw, pws, w = [], [], [], [], [], [], [], [], []
    cols=['time', 'rh', 'T', 'Td', 'a', 'x', 'pw', 'pws', 'h']
    ai = (0, 21, 33, 45, 69, 84, 123, 139, 153)
    ei = (17, 27, 39, 51, 75, 91, 131, 147, 160)
    f = open(file, 'r')
    for line in f:
        n = 0
        for var in (time, rh, t, td, a, x, pw, pws, w):
            var.append(line[ai[n]:ei[n]])
            n += 1
    df = pd.DataFrame({cols[0]: time, cols[1]: rh, cols[2]: t, cols[3]: td,
                       cols[4]: a, cols[5]: x, cols[6]: pw, cols[7]: pws,
                       cols[8]: w},columns=cols)
    df = dataframe_to_numeric(df)
    
    #Conversion into SI units
    df['T'] = df['T'] + 273.15 #temperature  °C -> K
    df['Td'] = df['Td'] + 273.15 #dewpoint  °C -> K
    df['a'] = df['a'] / 1000 #absolute humidity  g/m^3 -> Kg/m^3
    df['x'] = df['x'] / 1000 #mixing ratio  g/Kg -> Kg/Kg
    df['pw'] = df['pw'] * 100 #water vapor pressure  hPa -> Pa
    df['pws'] = df['pws'] * 100 #water vapor pressure  hPa -> Pa
    
    return(df)

def load_novatel(file, delim=';|,|\t|\*', decim='.', leap=37):
    """
    Reads in novatel data logged with 'hgpstools' by Hauke Daempfling
    Input:
        file - input file
        skipr - how many header lines to skip
        delim - delimiter of the logging file
        decim - decimal sign
        leap - leap seconds between UTC and TAI
    Output:
        df - DataFrame with selected columns relevant for wind vector
                calculation
    """
    #+++CONSTANTS+++
    gps_init = 315964800 #begin of GPS time in UNIX Epoch
    week_sec = 604800 #No. of seconds of one week
    smpl_rt = 125 #internal sample rate of the IMU, 1/s
    
    #+++PROGRAMME+++
    print('Loading Novatel data...')
    t = skiprow_column_finder(file, sep=delim)
    df = pd.read_csv(file, skiprows=t[0], sep=delim, header=None,
                     engine='python', names=range(t[1]))
    idx = (df.loc[df.ix[:,1].str.contains('%IMURATECORRIMUSA'),1].index.values,
              df.loc[df.ix[:,1].str.contains('%IMURATEPVASA'),1].index.values)    
    cols0 = ['time', 'gnss_time_1', 'gnss_time_2', 'theta_dot', 'psi_dot',
             'x_ddot', 'y_ddot', 'z_ddot']
    cols1 = ['time', 'gnss_time_1', 'gnss_time_2', 'lat', 'lon', 'height',
             'u_p', 'v_p', 'w_p', 'psi', 'theta', 'phi', 'ins_state']
    #NOTE: to convert 'm/s/sample' into m/s, multiply the value with the 
    #sampling-rate (NOT logging-rate) in Hz of the used IMU. In asynchronous
    #logging modes, the used SPAN-IGM-S1 IMU gets sampled at smpl_rt=125 Hz.
    df0 = pd.DataFrame({
                        cols0[0]: df.ix[idx[0],0].values, #server timestamp
                        cols0[1]: gps_init+ #gnss timestamps
                            pd.to_numeric(df.ix[idx[0],2].values)*week_sec+
                            df.ix[idx[0],3].values-leap+19,
                        cols0[2]: gps_init+df.ix[idx[0],4].values*week_sec+
                            pd.to_numeric(df.ix[idx[0],5].values)-leap+19,
                            #theta_dot, rad/s/sample:
                        cols0[3]: pd.to_numeric(
                                df.ix[idx[0],6].values)*smpl_rt*(-1),
                            #psi_dot, rad/s/sample:
                        cols0[4]: pd.to_numeric(
                                df.ix[idx[0],8].values)*smpl_rt*(-1),
                            #x_ddot, m/s/sample:
                        cols0[5]: pd.to_numeric(
                                df.ix[idx[0],9].values)*smpl_rt,
                            #y_ddot, m/s/sample:
                        cols0[6]: pd.to_numeric(
                                df.ix[idx[0],10].values)*smpl_rt,
                            #z_ddot, m/s/sample:
                        cols0[7]: pd.to_numeric(
                                df.ix[idx[0],11].values)*smpl_rt
                        },
                        columns=cols0
                        )
    df1 = pd.DataFrame({
                        cols1[0]: df.ix[idx[1],0].values, #server timestamp
                        cols1[1]: gps_init+ #gnss timestamps
                            pd.to_numeric(df.ix[idx[1],2].values)*week_sec+
                            df.ix[idx[1],3].values-leap+19,
                        cols1[2]: gps_init+df.ix[idx[1],4].values*week_sec+
                            pd.to_numeric(df.ix[idx[1],5].values)-leap+19,
                        cols1[3]: df.ix[idx[1],6].values, #lat, deg
                        cols1[4]: df.ix[idx[1],7].values, #lon, deg
                        cols1[5]: df.ix[idx[1],8].values, #height, m
                        cols1[6]: df.ix[idx[1],10].values, #u, m/s
                        cols1[7]: df.ix[idx[1],9].values, #v, m/s
                        cols1[8]: df.ix[idx[1],11].values, #w, m/s
                        cols1[9]: df.ix[idx[1],14].values, #psi, deg
                        cols1[10]: df.ix[idx[1],13].values, #theta, deg
                        cols1[11]: df.ix[idx[1],12].values, #phi, deg
                        cols1[12]: df.ix[idx[1],15].values #ins_state
                        },
                       columns=cols1
                       )
    #transfer inertial solution status ('ins_state' column in nova1
    #dataframe) into a numeric value and drop data without complete alignment:
    ins_sol = {'INS_INACTIVE': 0., 'INS_ALIGNING': 1., 'INS_HIGH_VARIANCE': 2.,
               'INS_SOLUTION_GOOD': 3., 'INS_SOLUTION_FREE': 6., 
               'INS_ALIGNMENT_COMPLETE': 7., 'DETERMINING_ORIENTIATION': 8., 
               'WAITING_INITIALPOS': 9.}
    for i in ins_sol:
        df1['ins_state'][df1['ins_state'] == i] = ins_sol[i]
    df1['ins_state'] = pd.to_numeric(df1['ins_state'])
    #convert all columns to float objects:
    for i in [df0, df1]:
        i = dataframe_to_numeric(i)
    #if no position has been found, the IMU's GNSS time is set to the begin of
    #GPS time (06-01-1980, 0000z) --> filter dates with timestamps older than
    #about 2015ish:        
    thrs_time = 1420066800 #Thu 1. Jan 00:00:00 CET 2015
    df1 = df1[(df1['gnss_time_1'] > thrs_time) &
       (df1['ins_state'].isin([2,3,7]))]
    df0 = df0[(df0['gnss_time_1'] > thrs_time) &
               (df0['gnss_time_1'] > float(df1['gnss_time_1'][:1])) &
               (df0['gnss_time_1'] < float(df1['gnss_time_1'][-1:]))]
    
    #convert to SI units (deg -> rad)
    for var in ['psi', 'phi', 'theta']:
        df1[var] = df1[var] * np.pi / 180
    
    return(df0,df1)

def load_usb1608(file):
    """
    reads in raw data from USB1608 stream
    TODO: return multiple channel
    Input:
        file - file string
    Ouput:
        data - data from file
    """
    #Kanäle werden abgescannt, dann Zeitstempel generiert und
    # im Log unterhalb des Zeitstempels geprintet
    
    def make_df(time, vals, smpl):
        """
        Generates pandas DataFrame after scanning of USB1608 Logger is
            completed.
        Input:
            time - timestamps
            vals - values
            smpl - number of samples in one bin
        Output:
            df - pd.DataFrame
        """
        #Create Timestamps between sample packets
        #first: throw away first packet of values, since the time difference
        #between first and last sample in the first packet cannot be known
        vals = vals[smpl[0]:]
        time_new = []
        for i in range(0,len(time)-1):
            time_diff = (time[i+1]-time[i])/smpl[i+1]
            for j in range(0,smpl[0]):
                time_new.append(time[i]+time_diff+j*time_diff)
        cols = ['time', 'V']
        df = pd.DataFrame({cols[0]: time_new, cols[1]: vals},
                           columns=cols)
        f.close()

        return(df)
    
    print('Loading USB1608 data...')
    f = open(file, 'r')
    time, vals, smpl = [], [], []
    while True:
        buffer = f.readline()
        if buffer == '': #EOF reached
            try:
                return(make_df(time, vals, smpl))
            except:
                f.close()
                print('WARNING: File ' + file +
                      ' is empty. Nothing to return.')
                return(None)
        if buffer[0:4] == 'time': #if a timestamp has been found
            try:
                buffer_time = float(buffer[6:23]) #get the timestamp
                buffer_vals = [] #empty values buffer
                buffer = f.readline() #read in next line
                buffer_samples = int(buffer[22:24]) #read number of samples
                n = 0
                while n < buffer_samples:
                    buffer = f.readline()
                    if buffer == '': #EOF reached
                        return(make_df(time, vals, smpl))
                    elif buffer[0:4] == 'Scan':
                        buffer_vals.append(float(buffer[46:54]))
                    else:
                        n = buffer_samples #exit the loop
                    n += 1
                if len(buffer_vals) == buffer_samples:
                    [vals.append(i) for i in buffer_vals]
                    time.append(buffer_time)
                    smpl.append(buffer_samples)
            except:
                pass
    return(None)    
